README.md
# Jargon2 Backends: Argon2 implementations for the Jargon2 API 

[![Build Status](https://travis-ci.org/kosprov/jargon2-backends.svg?branch=master)](https://travis-ci.org/kosprov/jargon2-backends)
[![Coverity Scan Build Status](https://scan.coverity.com/projects/14708/badge.svg)](https://scan.coverity.com/projects/kosprov-jargon2-backends)
[![Maven metadata URI](https://img.shields.io/maven-metadata/v/http/central.maven.org/maven2/com/kosprov/jargon2/jargon2-native-ri-backend/maven-metadata.xml.svg)](https://search.maven.org/#search%7Cga%7C1%7Cg%3A%22com.kosprov.jargon2%22%20AND%20a%3A%22jargon2-native-ri-backend%22)
[![Quality Gate](https://sonarcloud.io/api/project_badges/measure?project=com.kosprov.jargon2%3Ajargon2-backends&metric=alert_status)](https://sonarcloud.io/dashboard/index/com.kosprov.jargon2:jargon2-backends)
[![Security Rating](https://sonarcloud.io/api/project_badges/measure?project=com.kosprov.jargon2%3Ajargon2-backends&metric=security_rating)](https://sonarcloud.io/dashboard/index/com.kosprov.jargon2:jargon2-backends)
[![Dependency Status](https://www.versioneye.com/user/projects/5af3e85c0fb24f0e5baabe55/badge.svg)](https://www.versioneye.com/user/projects/5af3e85c0fb24f0e5baabe55)
[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](/LICENSE)

This repository aims to be a collection of `com.kosprov.jargon2.spi.Jargon2Backend` SPI implementations, ready to plugged into the Jargon2 API. Artifacts contain all service provider metadata and can be used by simply adding them into the runtime classpath.

## Security considerations

This section summarizes any security considerations that come with the use of this library. Make sure you evaluate them before choosing to use any of the Jargon2 backends provided here and visit this section regularly for any updates.

| Item |  Description |
| ---  | --- |
| Default backend native library can be bypassed | If you're using the default backend (`jargon2-native-ri-backend`), the shared library it binds to can be overridden by defining one of `-Djna.boot.library.path`, `-Djna.library.path` and `-Djna.nosys` system properties. See [Hardening your environment](#hardening-your-environment) for more details. |


## The default backend

Currently, there is only one implementation named `jargon2-native-ri-backend` that wraps the [Argon2 reference implementation](https://github.com/P-H-C/phc-winner-argon2 "Argon2 reference implementation repository"). It's unique characteristic is that it binds directly to the low-level API of the C code. This allows for two distinctive features of the high-level Jargon2 API:

- Ability to set memory lanes and threads independently
- Leverage Argon2 RI API for keyed-hashing and additional authentication data

### Usage

Simply add this dependency:

```xml
<dependency>
    <groupId>com.kosprov.jargon2</groupId>
    <artifactId>jargon2-native-ri-backend</artifactId>
    <version>1.1.1</version>
    <scope>runtime</scope>
</dependency>
```

`jargon2-native-ri-backend` contains `META-INF/services/com.kosprov.jargon2.spi.Jargon2Backend` metadata and is automatically discovered by Jargon2's discovery process. No build time dependency is necessary, so it's recommended to keep `scope` to `runtime`. 

### Pre-packaged binaries

To make adopting Jargon2 as easy as possible, `jargon2-native-ri-backend` has a transitive dependency to `jargon2-native-ri-binaries-generic`, an artifact that contains binaries of the reference implementation. They are available for Windows, Linux and macOS (x86-64 only), so it should work as-is on most systems.

Release `1.1.1` of `jargon2-native-ri-binaries-generic` contains binaries built from Argon2 release [20171227](https://github.com/P-H-C/phc-winner-argon2/releases/tag/20171227 "Argon2 RI release 20171227").

### Using different binaries

There are at least three reasons why one would need to use different binaries than those included in `jargon2-native-ri-binaries-generic`:

- Custom-built for a particular x86 micro-architecture
    
    The reference implementation contains a number of optimizations on the low-level algorithms of Argon2 and Blake2b that utilize SIMD instructions on modern processors. Expect a significant performance boost just by recompiling the C code for your particular CPU type. The gains are bigger if you're hashing with large memory and time costs.

- Different architecture

    Binaries are available only for the x86 architecture, so different processor architectures would need their own binaries.
    
- Security patches

    If Argon2 RI releases security patches, you would always have the option to recompile and switch to the patched binaries.

To change the binaries you have two options:

- Build Argon2 RI and install it as a system library

    `jargon2-native-ri-backend` uses [JNA](https://github.com/java-native-access/jna) to dynamically invoke native code. JNA searches for system libraries first, so installing on `/usr/lib/libargon2.so` will take precedence over the classpath binaries. You can change the search location by setting the `-Djna.library.path` property.

- Tweak Maven dependencies

    If installing native libraries on the host OS is not very convenient, you can package your binaries in a jar and add that to your application. Don't forget to exclude the transitive dependency to `jargon2-native-ri-binaries-generic`.
    
    ```xml
    <dependency>
        <groupId>com.kosprov.jargon2</groupId>
        <artifactId>jargon2-native-ri-backend</artifactId>
        <version>1.1.1</version>
        <scope>runtime</scope>
        <exclusions>
            <!-- exclude transitive dependency to generic binaries -->
            <exclusion>
                <groupId>com.kosprov.jargon2</groupId>
                <artifactId>jargon2-native-ri-binaries-generic</artifactId>
            </exclusion>
        </exclusions>
    </dependency>
    <!-- add dependency to optimized binaries -->
    <dependency>
        <groupId>com.mycompany.jargon2</groupId>
        <artifactId>my-argon2-optimized-binaries</artifactId>
        <version>1.0.0</version>
        <scope>runtime</scope>
    </dependency>
    ```
    Have a look at `jargon2-native-ri-binaries-generic` to see the folder structure required by JNA.
    
If you're having doubts on which Argon2 binaries are loaded, start the JVM with `-Djna.debug_load=true`.

### Hardening your environment

JNA searches for libraries in locations that can be can be controlled with `-Djna.boot.library.path`, `-Djna.library.path` and `-Djna.nosys` system properties. Keep your security engineers alerted and have them scan or change-detect for improper use of these properties. Make sure they protect them as they would protect JAAS login module or security manager system properties and configuration files. Changing those system properties so that a malicius native library gets loaded, will leak all your user's passwords.

If you are using SELinux, loading the native library from the classpath (as `jargon2-native-ri-binaries-generic` does) would not work. You would have to install the library in an accessible location. You can also define `-Djna.nounpack=true` to make sure the library is never unpacked from the classpath.

CHANGELOG.md
# Change Log

## v1.1.1

Release date: May 10, 2018

- Minor improvements in code and build script (backwards compatible)

## v1.1.0

Release date: Jan 06, 2017

- Updated Argon2 reference implementation binaries to release 20171227
- Minor fixes on code quality (backwards compatible)

## v1.0.1

Release date: Dec 30, 2017

- Minor fixes on code quality (backwards compatible)

## v1.0.0

Release date: Dec 29, 2017

- Initial version

jargon2-native-ri-backend/src/test/java/com/kosprov/jargon2/nativeri/backend/TestUtils.java
package com.kosprov.jargon2.nativeri.backend;

import com.kosprov.jargon2.spi.Jargon2Backend;
import org.apache.commons.codec.binary.Hex;

import static com.kosprov.jargon2.api.Jargon2.Type;
import static com.kosprov.jargon2.api.Jargon2.Version;
import static org.junit.Assert.*;

class TestUtils {

    static void hashTest(Type type, Version version, int timeCost, int memoryCost, int lanes, int threads, int hashLength, byte[] password, byte[] salt, byte[] secret, byte[] ad, String rawHashHex) {
        Jargon2Backend backend = new NativeRiJargon2Backend();

        byte[] calculatedRawHashBytes = backend.rawHash(type, version, memoryCost, timeCost, lanes, threads, hashLength, secret, ad, salt, password, null);
        String calculatedRawHashHex = Hex.encodeHexString(calculatedRawHashBytes);
        assertTrue(rawHashHex.equalsIgnoreCase(calculatedRawHashHex));
    }

    static void hashTest(Type type, Version version, int timeCost, int memoryCost, int lanes, int threads, String password, String salt, String rawHashHex, String encodedHash) {
        Jargon2Backend backend = new NativeRiJargon2Backend();

        byte[] saltBytes = salt.getBytes();
        byte[] passwordBytes = password.getBytes();
        String calculatedEncodedHash = backend.encodedHash(type, version, memoryCost, timeCost, lanes, threads, 32, null, null, saltBytes, passwordBytes, null);
        assertEquals(encodedHash, calculatedEncodedHash);

        boolean encodedMatched = backend.verifyEncoded(encodedHash, threads, null, null, passwordBytes, null);
        assertTrue(encodedMatched);

        byte[] calculatedRawHashBytes = backend.rawHash(type, version, memoryCost, timeCost, lanes, threads, 32, null, null, saltBytes, passwordBytes, null);
        String calculatedRawHashHex = Hex.encodeHexString(calculatedRawHashBytes);
        assertTrue(rawHashHex.equalsIgnoreCase(calculatedRawHashHex));

        boolean rawMatched = backend.verifyRaw(type, version, memoryCost, timeCost, lanes, threads, calculatedRawHashBytes, null, null, saltBytes, passwordBytes, null);
        assertTrue(rawMatched);
    }

    static void verify(String encodedHash, int threads, String password) {
        Jargon2Backend backend = new NativeRiJargon2Backend();

        boolean matches = backend.verifyEncoded(encodedHash, threads, null, null, password.getBytes(), null);
        assertTrue(matches);
    }

    static void hash(Type type, Version version, int timeCost, int memoryCost, int lanes, int threads, String password, String salt) {
        Jargon2Backend backend = new NativeRiJargon2Backend();

        byte[] hash = backend.rawHash(type, version, memoryCost, timeCost, lanes, threads, 32, null, null, salt.getBytes(), password.getBytes(), null);
        assertNotNull(hash);
        assertTrue(hash.length == 32);
    }
}


jargon2-native-ri-backend/src/test-long-running/java/com/kosprov/jargon2/nativeri/backend/ReferenceImplementationLongRunningTest.java
package com.kosprov.jargon2.nativeri.backend;

import org.junit.Test;

import static com.kosprov.jargon2.api.Jargon2.Type.ARGON2i;
import static com.kosprov.jargon2.api.Jargon2.Version.V10;
import static com.kosprov.jargon2.api.Jargon2.Version.V13;
import static com.kosprov.jargon2.nativeri.backend.TestUtils.hashTest;

/**
 * Test vectors from C reference implementation
 */
public class ReferenceImplementationLongRunningTest {

    /**
     * Large RAM test vectors from https://github.com/P-H-C/phc-winner-argon2/blob/master/src/test.c
     *
     * @throws Exception
     */
    @Test
    public void testWithReferenceImplTestVectorsLongRunning() throws Exception {

        hashTest(ARGON2i, V10, 2, (int) Math.pow(2, 20), 1, 1,
                "password", "somesalt",
                "9690ec55d28d3ed32562f2e73ea62b02b018757643a2ae6e79528459de8106e9",
                "$argon2i$m=1048576,t=2,p=1$c29tZXNhbHQ$lpDsVdKNPtMlYvLnPqYrArAYdXZDoq5ueVKEWd6BBuk");

        hashTest(ARGON2i, V10, 2, (int) Math.pow(2, 18), 1, 1,
                "password", "somesalt",
                "3e689aaa3d28a77cf2bc72a51ac53166761751182f1ee292e3f677a7da4c2467",
                "$argon2i$m=262144,t=2,p=1$c29tZXNhbHQ$Pmiaqj0op3zyvHKlGsUxZnYXURgvHuKS4/Z3p9pMJGc");

        hashTest(ARGON2i, V13, 2, (int) Math.pow(2, 20), 1, 1,
                "password", "somesalt",
                "d1587aca0922c3b5d6a83edab31bee3c4ebaef342ed6127a55d19b2351ad1f41",
                "$argon2i$v=19$m=1048576,t=2,p=1$c29tZXNhbHQ$0Vh6ygkiw7XWqD7asxvuPE667zQu1hJ6VdGbI1GtH0E");

        hashTest(ARGON2i, V13, 2, (int) Math.pow(2, 18), 1, 1,
                "password", "somesalt",
                "296dbae80b807cdceaad44ae741b506f14db0959267b183b118f9b24229bc7cb",
                "$argon2i$v=19$m=262144,t=2,p=1$c29tZXNhbHQ$KW266AuAfNzqrUSudBtQbxTbCVkmexg7EY+bJCKbx8s");

    }
}

jargon2-native-ri-backend/src/main/java/com/kosprov/jargon2/nativeri/backend/NativeRiJargon2Backend.java
package com.kosprov.jargon2.nativeri.backend;

import argon2.Argon2Library;
import argon2.Argon2_Context;
import com.kosprov.jargon2.spi.Jargon2Backend;
import com.kosprov.jargon2.spi.Jargon2BackendException;
import com.sun.jna.Memory;

import java.util.Arrays;
import java.util.Map;

import static argon2.Argon2Library.*;
import static argon2.Argon2Library.Argon2_ErrorCodes.ARGON2_OK;
import static argon2.Argon2Library.Argon2_type.*;
import static argon2.Argon2Library.Argon2_version.ARGON2_VERSION_10;
import static argon2.Argon2Library.Argon2_version.ARGON2_VERSION_13;
import static com.kosprov.jargon2.api.Jargon2.Type;
import static com.kosprov.jargon2.api.Jargon2.Version;

public class NativeRiJargon2Backend implements Jargon2Backend {

    @Override
    public byte[] rawHash(Type type, Version version, int memoryCost, int timeCost, int lanes, int threads, int hashLength, byte[] secret, byte[] ad, byte[] salt, byte[] password, Map<String, Object> options) {

        if (hashLength < ARGON2_MIN_OUTLEN) {
            throw new Jargon2BackendException("Hash length must be greater or equal to " + ARGON2_MIN_OUTLEN);
        }

        if (salt == null || salt.length < ARGON2_MIN_SALT_LENGTH) {
            throw new Jargon2BackendException("Salt must not be null and its length must be greater or equal to " + ARGON2_MIN_SALT_LENGTH);
        }

        if (password == null || password.length == 0) {
            throw new Jargon2BackendException("Password must not be null or empty");
        }

        if (lanes < ARGON2_MIN_LANES) {
            throw new Jargon2BackendException("Lanes must be greater or equal to " + ARGON2_MIN_LANES);
        }

        if (lanes > ARGON2_MAX_LANES) {
            throw new Jargon2BackendException("Lanes must be less or equal to " + ARGON2_MAX_LANES);
        }

        if (threads < ARGON2_MIN_THREADS) {
            throw new Jargon2BackendException("Threads must be greater or equal to " + ARGON2_MIN_THREADS);
        }

        if (threads > ARGON2_MAX_THREADS) {
            throw new Jargon2BackendException("Threads must be less or equal to " + ARGON2_MAX_THREADS);
        }

        if (threads > lanes) {
            threads = lanes;
        }

        if (memoryCost < ARGON2_MIN_MEMORY) {
            throw new Jargon2BackendException("Memory cost must be greater or equal to " + ARGON2_MIN_MEMORY);
        }

        if (memoryCost < 2 * ARGON2_SYNC_POINTS * lanes) {
            throw new Jargon2BackendException("Memory cost must be greater or equal to " + (2 * ARGON2_SYNC_POINTS) + " * lanes");
        }

        if (timeCost < ARGON2_MIN_TIME) {
            throw new Jargon2BackendException("Time cost must be greater or equal to " + ARGON2_MIN_TIME);
        }


        DisposableMemory passwordMemory = null;
        DisposableMemory saltMemory = null;
        DisposableMemory secretMemory = null;
        DisposableMemory adMemory = null;
        DisposableMemory outputMemory = null;

        try {
            passwordMemory = copyToMemory(password);
            saltMemory = copyToMemory(salt);
            secretMemory = copyToMemory(secret);
            adMemory = copyToMemory(ad);
            outputMemory = createMemory(hashLength);

            Argon2_Context ctx = new Argon2_Context.ByReference();
            ctx.out = outputMemory;
            ctx.outlen = hashLength;
            ctx.pwd = passwordMemory;
            ctx.pwdlen = password.length;
            ctx.salt = saltMemory;
            ctx.saltlen = salt.length;
            ctx.secret = secretMemory;
            ctx.secretlen = secret != null ? secret.length : 0;
            ctx.ad = adMemory;
            ctx.adlen = ad != null ? ad.length : 0;
            ctx.t_cost = timeCost;
            ctx.m_cost = memoryCost;
            ctx.lanes = lanes;
            ctx.threads = threads;
            ctx.version = convertVersion(version);
            ctx.allocate_cbk = null;
            ctx.free_cbk = null;
            ctx.flags = ARGON2_DEFAULT_FLAGS;

            int argon2Type = convertType(type);

            int status = Argon2Library.INSTANCE.argon2_ctx(ctx, argon2Type);

            if (status != ARGON2_OK) {
                String errorMessage = Argon2Library.INSTANCE.argon2_error_message(status);
                throw new NativeInvocationJargon2BackendException(status, errorMessage);
            }

            byte[] hash = new byte[hashLength];
            outputMemory.read(0, hash, 0, hashLength);

            return hash;
        } finally {
            if (passwordMemory != null) {
                passwordMemory.clear(password.length);
                passwordMemory.dispose();
            }
            if (saltMemory != null) {
                saltMemory.dispose();
            }
            if (secretMemory != null) {
                secretMemory.clear(secret.length);
                secretMemory.dispose();
            }
            if (adMemory != null) {
                adMemory.dispose();
            }
            if (outputMemory != null) {
                outputMemory.clear(hashLength);
                outputMemory.dispose();
            }
        }
    }

    @Override
    public String encodedHash(Type type, Version version, int memoryCost, int timeCost, int lanes, int threads, int hashLength, byte[] secret, byte[] ad, byte[] salt, byte[] password, Map<String, Object> options) {
        byte[] rawHash = rawHash(type, version, memoryCost, timeCost, lanes, threads, hashLength, secret, ad, salt, password, options);
        try {
            return encodeString(type, version, timeCost, memoryCost, lanes, salt, rawHash);
        } catch (Exception e) {
            throw new Jargon2BackendException("Failed to encode hash", e);
        } finally {
            Arrays.fill(rawHash, (byte) 0x00);
        }
    }

    @Override
    public boolean verifyEncoded(String encodedHash, int threads, byte[] secret, byte[] ad, byte[] password, Map<String, Object> options) {
        if (encodedHash == null) {
            throw new Jargon2BackendException("Encoded hash cannot be null");
        }

        if ("".equals(encodedHash.trim())) {
            throw new Jargon2BackendException("Encoded hash cannot be empty");
        }

        DecodedHash decoded = decodeString(encodedHash);
        int lanes = decoded.parallelism;
        if (threads == -1) {
            threads = decoded.parallelism;
        }
        return verifyRaw(decoded.type, decoded.version, decoded.memoryCost, decoded.timeCost, lanes, threads, decoded.hash, secret, ad, decoded.salt, password, options);
    }

    @Override
    public boolean verifyRaw(Type type, Version version, int memoryCost, int timeCost, int lanes, int threads, byte[] rawHash, byte[] secret, byte[] ad, byte[] salt, byte[] password, Map<String, Object> options) {
        byte[] newHash = rawHash(type, version, memoryCost, timeCost, lanes, threads, rawHash.length, secret, ad, salt, password, options);
        return Arrays.equals(rawHash, newHash);
    }

    private DecodedHash decodeString(String encodedHash) {
        DecodedHash decoded = new DecodedHash();

        String[] parts = split(encodedHash, '$', 5);

        if (parts.length != 4 && parts.length != 5) {
            throw new Jargon2BackendException("Encoded hash is not properly formatted");
        }

        int typeIndex = 0;
        int versionIndex = 1;
        int optionsIndex = parts.length == 4 ? 1 : 2;
        int saltIndex = parts.length == 4 ? 2 : 3;
        int hashIndex = parts.length == 4 ? 3 : 4;

        String typeOpt = parts[typeIndex];
        decoded.type = convertType(typeOpt);

        if (parts.length == 5) {
            String versionOpt = parts[versionIndex];
            if (!versionOpt.startsWith("v=") || versionOpt.length() < 4) {
                throw new Jargon2BackendException("Encoded hash is not properly formatted");
            }
            decoded.version = convertVersion(versionOpt.substring(2));
        } else {
            decoded.version = Version.V10;
        }

        parseOptions(parts[optionsIndex], decoded);

        try {
            decoded.salt = base64decode(parts[saltIndex]);
        } catch (Exception e) {
            throw new Jargon2BackendException("Could not decode salt", e);
        }

        try {
            decoded.hash = base64decode(parts[hashIndex]);
        } catch (Exception e) {
            throw new Jargon2BackendException("Could not decode hash", e);
        }

        return decoded;
    }

    private void parseOptions(String options, DecodedHash decoded) {
        String[] opts = split(options, ',', 3);
        if (opts.length != 3) {
            throw new Jargon2BackendException("Wrong number of hashing options");
        }

        String memoryCostOption = opts[0];
        if (!memoryCostOption.startsWith("m=") || memoryCostOption.length() < 3) {
            throw new Jargon2BackendException("Wrong memory cost option");
        }
        try {
            decoded.memoryCost = Integer.parseInt(memoryCostOption.substring(2));
        } catch (Exception e) {
            throw new Jargon2BackendException("Memory cost option is invalid");
        }

        String timeCostOption = opts[1];
        if (!timeCostOption.startsWith("t=") || timeCostOption.length() < 3) {
            throw new Jargon2BackendException("Wrong time cost option");
        }
        try {
            decoded.timeCost = Integer.parseInt(timeCostOption.substring(2));
        } catch (Exception e) {
            throw new Jargon2BackendException("Time cost option is invalid");
        }

        String parallelismOption = opts[2];
        if (!parallelismOption.startsWith("p=") || parallelismOption.length() < 3) {
            throw new Jargon2BackendException("Wrong parallelism option");
        }
        try {
            decoded.parallelism = Integer.parseInt(parallelismOption.substring(2));
        } catch (Exception e) {
            throw new Jargon2BackendException("Parallelism option is invalid");
        }
    }

    private String encodeString(Type type, Version version, int timeCost, int memoryCost, int lanes, byte[] salt, byte[] hash) {
        StringBuilder sb = new StringBuilder();
        sb.append('$').append(type.getValue());
        if (version.getValue() > Version.V10.getValue()) {
            sb.append('$').append("v=").append(version.getValue());
        }
        sb.append('$').append("m=").append(memoryCost).append(",t=").append(timeCost).append(",p=").append(lanes);
        sb.append('$').append(base64encode(salt));
        sb.append('$').append(base64encode(hash));
        return sb.toString();
    }

    private DisposableMemory createMemory(int length) {
        return new DisposableMemory(length);
    }

    private DisposableMemory copyToMemory(byte[] arr) {
        if (arr == null) return null;
        DisposableMemory m = new DisposableMemory(arr.length);
        m.write(0, arr, 0, arr.length);
        return m;
    }

    private int convertVersion(Version version) {
        if (Version.V13.equals(version)) {
            return ARGON2_VERSION_13;
        } else if (Version.V10.equals(version)) {
            return ARGON2_VERSION_10;
        } else {
            throw new Jargon2BackendException("Null or unsupported version detected: " + version);
        }
    }

    private Version convertVersion(String version) {
        try {
            int v = Integer.parseInt(version);
            if (Version.V13.getValue() == v) {
                return Version.V13;
            } else if (Version.V10.getValue() == v) {
                return Version.V10;
            } else {
                throw new Jargon2BackendException("Invalid version number. Check encoded hash.");
            }
        } catch (NumberFormatException e) {
            throw new Jargon2BackendException("Non-numeric version. Check encoded hash.");
        }
    }

    private int convertType(Type type) {
        if (Type.ARGON2d.equals(type)) {
            return Argon2_d;
        } else if (Type.ARGON2i.equals(type)) {
            return Argon2_i;
        } else if (Type.ARGON2id.equals(type)) {
            return Argon2_id;
        } else {
            throw new Jargon2BackendException("Null or unsupported type detected: " + type);
        }
    }

    private Type convertType(String type) {
        if (Type.ARGON2d.getValue().equals(type)) {
            return Type.ARGON2d;
        } else if (Type.ARGON2i.getValue().equals(type)) {
            return Type.ARGON2i;
        } else if (Type.ARGON2id.getValue().equals(type)) {
            return Type.ARGON2id;
        } else {
            throw new Jargon2BackendException("Type did not decode properly. Check encoded hash.");
        }
    }

    static class DecodedHash {
        Type type;
        Version version;
        int memoryCost;
        int timeCost;
        int parallelism;
        byte[] salt;
        byte[] hash;
    }

    static class DisposableMemory extends Memory {
        DisposableMemory(long size) {
            super(size);
        }

        @Override
        public void dispose() {
            super.dispose();
        }
    }

    String[] split(String str, char delimiter, int maxTokens) {

        String[] tokens = new String[maxTokens];
        int from = 0;
        int i = 0;
        int len = str.length();

        while (from < len - 1) {
            if (i > tokens.length - 1) {
                break;
            }
            int delimIndex = str.indexOf(delimiter, from);
            if (delimIndex == -1) {
                if (from < len) {
                    tokens[i++] = str.substring(from, len);
                }
                break;
            }
            if (delimIndex > from) {
                tokens[i++] = str.substring(from, delimIndex);
            }
            from = delimIndex + 1;
        }

        if (i < tokens.length) {
            tokens = Arrays.copyOf(tokens, i);
        }

        return tokens;
    }

    private static final char[] encodeMapping = {
        'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P',
        'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', 'a', 'b', 'c', 'd', 'e', 'f',
        'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v',
        'w', 'x', 'y', 'z', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '+', '/'
    };

    char[] base64encode(byte[] data) {
        int base64Length = data.length / 3 * 4;
        int mod3 = data.length % 3;
        if (mod3 != 0) {
            base64Length += (mod3 + 1);
        }
        char[] base64 = new char[base64Length];
        int i = 0;
        int j = 0;

        while (i < data.length - 2) {
            base64[j++] = encodeMapping[data[i] >>> 2 & 0b00111111];
            base64[j++] = encodeMapping[(data[i] << 4 & 0b00110000) | (data[++i] >>> 4 & 0b00001111)];
            base64[j++] = encodeMapping[(data[i] << 2 & 0b00111100) | (data[++i] >>> 6 & 0b00000011)];
            base64[j++] = encodeMapping[data[i++] & 0b00111111];
        }
        if (i == data.length - 1) {
            base64[j++] = encodeMapping[data[i] >>> 2 & 0b00111111];
            base64[j] = encodeMapping[data[i] << 4 & 0b00110000];
        } else if (i == data.length - 2) {
            base64[j++] = encodeMapping[data[i] >>> 2 & 0b00111111];
            base64[j++] = encodeMapping[(data[i] << 4 & 0b00110000) | (data[++i] >>> 4 & 0b00001111)];
            base64[j] = encodeMapping[data[i] << 2 & 0b00111100];
        }

        return base64;
    }

    private static final byte[] decodeMapping = {
            -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
            -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
            -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 62, -1, -1, -1, 63,
            52, 53, 54, 55, 56, 57, 58, 59, 60, 61, -1, -1, -1, -1, -1, -1,
            -1,  0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14,
            15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, -1, -1, -1, -1, -1,
            -1, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40,
            41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, -1, -1, -1, -1, -1
    };

    private static byte decodeMapping(char c) {
        byte decoded;
        if (c > 127 || (decoded = decodeMapping[c]) == -1) {
            throw new Jargon2BackendException("Invalid character in base64 string");
        }
        return decoded;
    }

    byte[] base64decode(String encoded) {
        int outputLength = encoded.length() / 4 * 3;
        int mod4 = encoded.length() % 4;
        if (mod4 != 0) {
            outputLength += (mod4 - 1);
        }
        byte[] output = new byte[outputLength];

        char[] buf = new char[4];
        int i = 0;
        int j = 0;

        while (i < encoded.length() - 3) {
            encoded.getChars(i, i + 4, buf, 0);
            output[j++] = (byte) ((decodeMapping(buf[0]) << 2) | (decodeMapping(buf[1]) >>> 4));
            output[j++] = (byte) ((decodeMapping(buf[1]) << 4) | (decodeMapping(buf[2]) >>> 2));
            output[j++] = (byte) ((decodeMapping(buf[2]) << 6) | (decodeMapping(buf[3]) & 0xFF));
            i += 4;
        }

        if (i == encoded.length() - 2) {
            encoded.getChars(i, i + 2, buf, 0);
            output[j] = (byte) ((decodeMapping(buf[0]) << 2) | (decodeMapping(buf[1]) >>> 4));
        } else if (i == encoded.length() - 3) {
            encoded.getChars(i, i + 3, buf, 0);
            output[j++] = (byte) ((decodeMapping(buf[0]) << 2) | (decodeMapping(buf[1]) >>> 4));
            output[j] = (byte) ((decodeMapping(buf[1]) << 4) | (decodeMapping(buf[2]) >>> 2));
        } else if (i != encoded.length()) {
            throw new Jargon2BackendException("Wrong number of characters in base64 string");
        }

        return output;
    }
}
